## Problema 1

### EMV

Primero, construiremos el clasificador usando el estimador de máxima verosimilitud. 

_Asumimos que nuestro modelo sigue una distribución de probabilidad de Bernoulli(...)_

Comenzaremos por calcular, dada la tabla descrita en el problema, las probabilidades a priori siguientes:

- $P(JA=Joven)=\frac{6}{13}$
- $P(JA=Adulto)=\frac{7}{13}$

Ahora bien, sea X = (Pink Floyd, The Beatles, REM, Nirvana, Queen, Oasis), asumiendo que las variables son condicionalmente independientes, calculamos las probabilidades de verosimilitud para la clase Joven:

- $P(Pink Floyd = 1|JA = Joven) = \frac{6}{6}$
- $P(The Beatles = 1 | JA = Joven) = \frac{3}{6}$
- $P(REM = 1 | JA = Joven) = \frac{3}{6}$
- $P(Nirvana = 1 | JA = Joven) = \frac{2}{6}$
- $P(Queen = 1 | JA = Joven) = \frac{3}{6}$
- $P(Oasis= 1 | JA = Joven) = \frac{3}{6}$

Hacemos lo mismo para la clase Adulto:

- $P(Pink Floyd = 1 | JA = Adulto) = \frac{7}{7}$
- $P(The Beatles = 1 | JA = Adulto) = \frac{7}{7}$
- $P(REM = 1 | JA = Adulto) = \frac{4}{7}$
- $P(Nirvana = 1 | JA = Adulto) = \frac{3}{7}$
- $P(Queen = 1 | JA = Adulto) = \frac{5}{7}$
- $P(Oasis = 1 | JA = Adulto) = \frac{3}{7}$



Con los parámetros anteriores, clasificamos a cada uno de los vectores dados:

- $x_1 = (1, 1, 0, 1, 1, 0)$

$P(Joven|x_1)\ \alpha\ P(Joven) \times\ (P(PinkFloyd=1|Joven) \times P(TheBeatles=1|Joven) \times (1 - P(REM=1|Joven)) \times P(Nirvana=1|Joven) \times P(Queen=1|Joven) \times (1 - P(Oasis=1|Joven)))$

$$P(Joven|x_1)\ \alpha\ \frac{6}{13} \times\ (\frac{6}{6} \times \frac{3}{6} \times (1 - \frac{3}{6}) \times \frac{2}{6} \times \frac{3}{6} \times (1 - \frac{3}{6}))$$

In [20]:
(6.0/13)*(6.0/6)*(3.0/6)*(1 - (3.0/6))*(2.0/6)*(3.0/6)*(1-(3.0/6))

0.009615384615384616

$$P(Joven|x_1) = 0.0096$$

$P(Adulto|x_1)\ \alpha\ P(Adulto) \times\ (P(PinkFloyd=1|Adulto) \times P(TheBeatles=1|Adulto) \times (1 - P(REM=1|Adulto)) \times P(Nirvana=1|Adulto) \times P(Queen=1|Adulto) \times (1 - P(Oasis=1|Adulto)))$

$$P(Adulto|x_1)\ \alpha\ \frac{7}{13} \times\ (\frac{7}{7} \times \frac{7}{7} \times (1 - \frac{4}{7}) \times \frac{3}{7} \times \frac{5}{7} \times (1 - \frac{3}{7}))$$

In [24]:
(7.0/13)*(7.0/7)*(7.0/7)*(1 - (4.0/7))*(3.0/7)*(5.0/7)*(1-(3.0/7))

0.040367795469836286

$$P(Adulto|x_1) = 0.0404$$

Ya que $P(Adulto|x_1)\ >\ P(Joven|x_1)$, concluimos que el vector $x_1$ corresponde a la clase Adulto.

---

- $x_2 = (1,0,1,1,1,1)$

$$P(Joven|x_2)\ \alpha\ \frac{6}{13} \times\ (\frac{6}{6} \times (1 - \frac{3}{6}) \times \frac{3}{6} \times \frac{2}{6} \times \frac{3}{6} \times \frac{3}{6})$$



In [22]:
(6.0/13)*(6.0/6)*(1 - (3.0/6))*(3.0/6)*(2.0/6)*(3.0/6)*(3.0/6)

0.009615384615384616

$$P(Joven|x_2) = 0.0096$$

$$P(Adulto|x_2)\ \alpha\ \frac{7}{13} \times\ (\frac{7}{7} \times (1 - \frac{7}{7}) \times \frac{4}{7} \times \frac{3}{7} \times \frac{5}{7} \times \frac{3}{7})$$

$$P(Adulto|x_2) = 0$$

Ya que $P(Joven|x_2) > P(Adulto|x_2)$, concluimos que el vector $x_2$ corresponde a la clase Joven.

---

- $x_3 = (1, 1, 0, 0, 0, 0)$

$$P(Joven|x_3)\ \alpha\ \frac{6}{13} \times\ (\frac{6}{6} \times \frac{3}{6} \times (1-\frac{3}{6}) \times (1-\frac{2}{6}) \times (1-\frac{3}{6}) \times (1-\frac{3}{6}))$$

In [26]:
(6.0/13)*(6.0/6)*(3.0/6)*(1 - (3.0/6))*(1-(2.0/6))*(1-(3.0/6))*(1-(3.0/6))

0.019230769230769235

$$P(Joven|x_3) = 0.0192$$

$$P(Adulto|x_3)\ \alpha\ \frac{7}{13} \times\ (\frac{7}{7} \times \frac{7}{7} \times (1-\frac{4}{7}) \times (1-\frac{3}{7}) \times (1-\frac{5}{7}) \times (1-\frac{3}{7}))$$

In [30]:
(7.0/13)*(7.0/7)*(7.0/7)*(1 - (4.0/7))*(1-(3.0/7))*(1-(5.0/7))*(1-(3.0/7))

0.021529490917246017

$$P(Adulto|x_3) = 0.0215$$

Ya que $P(Adulto|x_3) > P(Joven|x_3)$, concluimos que el vector $x_3$ corresponde a la clase Adulto.

---

- $x_4 = (1, 1, 1, 1, 1, 1)$

$$P(Joven|x_4)\ \alpha\ \frac{6}{13} \times\ (\frac{6}{6} \times \frac{3}{6} \times \frac{3}{6} \times \frac{2}{6} \times \frac{3}{6} \times \frac{3}{6})$$

In [28]:
(6.0/13)*(6.0/6)*(3.0/6)*(3.0/6)*(2.0/6)*(3.0/6)*(3.0/6)

0.009615384615384616

$$P(Joven|x_4) = 0.0096$$

$$P(Adulto|x_4)\ \alpha\ \frac{7}{13} \times\ (\frac{7}{7} \times \frac{7}{7} \times \frac{4}{7} \times \frac{3}{7} \times \frac{5}{7} \times \frac{3}{7})$$

In [29]:
(7.0/13)*(7.0/7)*(7.0/7)*(4.0/7)*(3.0/7)*(5.0/7)*(3.0/7)

0.04036779546983627

$$P(Adulto|x_4) = 0.0404$$

Ya que $P(Adulto|x_4) > P(Joven|x_4)$, concluimos que el vector $x_4$ corresponde a la clase Adulto.

---

- $x_5 = (0, 1, 1, 1, 1, 1)$

$$P(Joven|x_5)\ \alpha\ \frac{6}{13} \times\ ((1-\frac{6}{6}) \times \frac{3}{6} \times \frac{3}{6} \times \frac{2}{6} \times \frac{3}{6} \times \frac{3}{6})$$

$$P(Joven|x_5) = 0$$

$$P(Adulto|x_5)\ \alpha\ \frac{7}{13} \times\ ((1-\frac{7}{7}) \times \frac{7}{7} \times \frac{4}{7} \times \frac{3}{7} \times \frac{5}{7} \times \frac{3}{7})$$

$$P(Adulto|x_5) = 0$$

Ya que $P(Joven|x_5) = P(Adulto|x_5)$ no es posible clasificar al vector $x_5$ usando el estimador de máxima verosimilitud planteado.

---

### MAP

A continuación, realizaremos la clasificación de los vectores anteriores usando un estimador de máximo a posteriori para calcular todos los parámetros de los atributos definidos por X y la clase C. Nuevamente asumiremos una distribución de probabilidad de Bernoulli, y para calcular los parámetros con MAP, usaremos usaremos:

$$q_{C} = \frac{N_C+ \alpha - 1}{N + \beta + \alpha - 2}$$

$$q_{x|C} = \frac{n_c(x)+ \alpha - 1}{N + \beta + \alpha - 2}$$

Donde asumiremos que:

- Para los parámetros de los atributos, $\alpha = 2$ y $\beta=|V|$ (donde |V| representa el número de valores que puede tomar el atributo en cuestión). En todos los casos $\beta = 2$.
- Para los parámetros de la clase, $\alpha = 2$ y $\beta=|C|$ (donde C representa el número de valores que puede tomar la clase en cuestión). En todos los casos $\beta = 2$.

Calculamos pues los parámetros de clase:

- $q_{Joven} = \frac{6+2-1}{13+2+2-2} = \frac{7}{15}$
- $q_{Adulto} = \frac{7+2-1}{13+2+2-2} = \frac{8}{15}$

Para la parte restante de este inciso, y en adelante, automatizaremos el cálculo de cada uno de los parámetros de clase y de atributos. Para hacerlo, importaremos la tabla mostrada en el problema a un arreglo que pueda ser manipulado por NumPy.



In [129]:
qj = 7.0/15
qa = 8.0/15

Para la parte restante de este inciso, y en adelante, automatizaremos el cálculo de cada uno de los parámetros de clase y de atributos. Para hacerlo, importaremos la tabla mostrada en el problema a un arreglo que pueda ser manipulado por NumPy.


In [34]:
import pandas

In [35]:
bandas = pandas.read_csv('bandas.csv')

In [119]:
bandas

Pink Floyd  The Beatles  REM  Nirvana  Queen  Oasis JA
0            1            0    0        1      1      1  J
1            1            1    0        1      1      0  J
2            1            1    1        0      0      1  J
3            1            0    1        0      0      1  J
4            1            0    0        0      1      0  J
5            1            1    1        0      0      0  J
6            1            1    0        0      1      1  A
7            1            1    1        0      0      1  A
8            1            1    1        1      1      0  A
9            1            1    1        0      1      0  A
10           1            1    1        0      1      1  A
11           1            1    0        1      1      0  A
12           1            1    0        1      0      0  A

Primero, dividimos a la tabla anterior en un arreglo "v" donde esté contenida la información de las bandas, y un vector "c" que tenga la información de las clases. Esto con la intención de poder manejar los datos más fácilmente en un programa.

In [118]:
b = bandas.values
v = b[:,:-1]
c = b[:,-1]

Ahora, para calcular los parámetros $q_{a|C}$, usamos la siguiente función, que toma como entradas el nombre de la clase en cuestión (_C_),  y los valores de $\alpha$ y $\beta$ definidos previamente.

In [110]:
def getParamsForClass(name, alpha, beta):
    n = 0
    nc = 0
    params = []
    for i in range(len(c)):
        if c[i] == name:
            n = n + 1
    for i in range(len(v[0])):
        for j in range(len(v)):
            if c[j] == name:
                nc = nc + v[j][i]
        params.append(1.0*(nc + alpha - 1)/(n + beta + alpha - 2))
        nc = 0

    return params

In [124]:
jparams = getParamsForClass('J', 2, 2)
jparams

[0.875, 0.5, 0.5, 0.375, 0.5, 0.5]

Esto es, para la clase Joven, tenemos:

- $P_{PinkFloyd|Joven}=0.875$
- $P_{TheBeatles|Joven}=0.5$
- $P_{REM|Joven}=0.5$
- $P_{Nirvana|Joven}=0.375$
- $P_{Queen|Joven}=0.5$
- $P_{Oasis|Joven}=0.5$

In [126]:
aparams = getParamsForClass('A', 2, 2)
aparams

[0.8888888888888888,
 0.8888888888888888,
 0.5555555555555556,
 0.4444444444444444,
 0.6666666666666666,
 0.4444444444444444]

Por su parte, para la clase Adulto, tenemos:

- $P_{PinkFloyd|Adulto}=0.8889$
- $P_{TheBeatles|Adulto}=0.8889$
- $P_{REM|Adulto}=0.5556$
- $P_{Nirvana|Adulto}=0.4444$
- $P_{Queen|Adulto}=0.6667$
- $P_{Oasis|Adulto}=0.4444$

Con los parámetros obtenidos, obtenemos las probabilidades de pertenencia a cada una de las clases para cada uno de los vectores de prueba, usando la siguiente función. Esta función toma como entradas el parámetro de la clase en cuestión ("cls"), el vector de prueba ("testvector") y los parámetros asociados a la clase ("params"), y regresa la probabilidad estimada $P(C|X)$.

In [135]:
def getProbabilities(cls, testvector, params):
    prob = 1
    for i in range(len(testvector)):
        if testvector[i] == 1:
            prob = prob * params[i]
        else:
            prob = prob * (1 - params[i])
    return cls * prob

Así pues, clasificamos a cada vector según la máxima probabilidad obtenida.

- $x_1 = (1,1,0,1,1,0)$

In [133]:
getProbabilities(qj, [1,1,0,1,1,0], jparams) 

0.0095703125

In [137]:
getProbabilities(qa, [1,1,0,1,1,0], aparams) 

0.030829386517035755

Tenemos entonces que:

$$P(Joven|x_1) = 0.0096$$
$$P(Adulto|x_1) = 0.0309$$

Por lo tanto, concluimos que $x_1$ pertenece a la clase Adulto.

---

- $x_2 = (1, 0, 1, 1, 1, 1)$

In [139]:
getProbabilities(qj, [1,0,1,1,1,1], jparams) 

0.0095703125

In [148]:
getProbabilities(qa, [1,0,1,1,1,1], aparams) 

0.003853673314629471

Tenemos entonces que :

$$P(Joven|x_2) = 0.0096$$
$$P(Adulto|x_2) = 0.0038$$

Por lo tanto, concluimos que $x_2$ pertenece a la clase Joven.

---

- $x_3 = (1, 1, 0, 0, 0, 0)$


In [143]:
getProbabilities(qj, [1,1,0,0,0,0], jparams) 

0.015950520833333332

In [145]:
getProbabilities(qa, [1,1,0,0,0,0], aparams) 

0.01926836657314735

Tenemos entonces que :

$$P(Joven|x_3) = 0.0159$$
$$P(Adulto|x_3) = 0.0192$$

Por lo tanto, concluimos que $x_3$ pertenece a la clase Adulto.

---

- $x_4 = (1, 1, 1, 1, 1, 1)$

In [146]:
getProbabilities(qj, [1,1,1,1,1,1], jparams) 

0.0095703125

In [149]:
getProbabilities(qa, [1,1,1,1,1,1], aparams) 

0.03082938651703575

Tenemos entonces que :

$$P(Joven|x_4) = 0.0096$$
$$P(Adulto|x_4) = 0.0308$$

Por lo tanto, concluimos que $x_4$ pertenece a la clase Adulto.

---

- $x_5 = (0, 1, 1, 1, 1, 1)$

In [154]:
getProbabilities(qj, [0,1,1,1,1,1], jparams) 

0.0013671875

In [153]:
getProbabilities(qa, [0,1,1,1,1,1], aparams) 

0.003853673314629471

Tenemos entonces que :

$$P(Joven|x_5) = 0.0014$$
$$P(Adulto|x_5) = 0.0039$$

Por lo tanto, concluimos que $x_5$ pertenece a la clase Adulto.